proposals/text/3925-mrkl-agent.md

```diff
@@ -0,0 +1,380 @@
+
+```
+I need to find out his age
+Action: Search
+Action Input: Jason Sudeikis age
+```
+
+This LLM response above gives us enough information to invoke a Search tool again along with the appropriate input, and we receive the response from the Search: 47 years. We add this response to the prompt history as the `Observation:` keyword.
+
+### Step 3:
+
+For the sake of brevity, let's not list the entire prompt again. The critical part to remember is that we append the output of step 2 to the prompt history we are creating as we step through each agent step. These so-called reasoning traces help agents "understand" what needs to be done in each successive step.
+The last part of the prompt is the following:
+```
+Thought: I need to find out his age
+Action: Search
+Action Input: Jason Sudeikis age
+Observation: 47 years
+Thought:
+```
+
+The LLM-generated response is:
+
+```
+I need to raise it to the 0.23 power
+Action: Calculator
+Action Input: 47^0.23
+```
+
+In this step, we invoke a new tool - The calculator with specified input. The calculator response is 2.4242784855673896
+We added the calculator response to the prompt history under the Observation keyword.
+
+
+### Step 4:
+
+Again, we append a calculator response and prompt to the prompt history once again. Let's not list the entire prompt, but the last few lines:
+```
+I need to raise it to the 0.23 power
+Action: Calculator
+Action Input: 47^0.23
+Observation: 2.4242784855673896
+Thought:
+```
+
+The LLM-generated response is:
+```
+I now know the final answer
+Final Answer: Jason Sudeikis, Olivia Wilde's boyfriend, is 47 years old and his age raised to the 0.23 power is 2.4242784855673896.
+```
+
+Using simple string parsing, we can detect that the mode in this step responded with the "Final Answer:" keyword just as we instructed, thus breaking out of the loop and completing the agent's task with a response returned to the agent's client.
+In the rare case that "Final Answer:" is not generated even after many iterations, we can break out of the loop based on a maximum number of iterations allowed. Thereby, we can prevent an infinite loop.
+
+## Agent Creation
+
+The Agent can be either created programmatically or loaded from a YAML file.
+In the following example, one tool is a node for searching the web. The other tool is a pipeline for doing calculations in python.
+
+**Example programmatic creation:**
+```python
+search = SerpAPIComponent(api_key=os.environ.get("SERPAPI_API_KEY"), name="Serp", inputs=["Query"])
+
+prompt_model=PromptModel(model_name_or_path="text-davinci-003", api_key=os.environ.get("OPENAI_API_KEY"))
+
+calculator = Pipeline()
+calculator.add_node(PromptNode(
+    model_name_or_path=prompt_model,
+    default_prompt_template=PromptTemplate(prompt_text="Write a simple python function that calculates..."),
+    output_variable="python_runtime_input")  # input
+calculator.add_node(PythonRuntime())  # actual calculator
```

Comment:
I would be careful with using something like the python runtime as the calculator. From what I understand this would allow the Agent to execute a string as python code which could pose a security risk. Since the construction of the string is controllable at run time it could potentially be abused. I do see the security risk being more pronounced for the `PythonRuntime()` node in isolation, but I think it is something to consider. 

I like the idea of using the SerpAPI to calculate simple expressions, or we could consider connecting to other external APIs like the [Wolfram Alpha API](https://products.wolframalpha.com/api) (e.g. the Instant Calculator API).

Reply:
Yes, PythonRuntime is quite powerful and could be easily misused. Same for a tool that could use command line. For the discussion what tools to implement we will have another proposal. Wolfram Alpha is a good idea if we have a use case that requires calculations. 👍

Reply:
```
runtime_globals = {**globals()}
exec(python_runtime_input, runtime_globals, None)
```

May I suggest using own globals and locals? Without unpacking like this. It will be more secure.